### PR TITLE
Adding .NET SDK for Azure Carbon Optimization.

### PR DESCRIPTION
**Pull request by destinyboyer**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

**Review by jsquire — DISMISSED** [review]
@destinyboyer:

Your account is missing the necessary write permissions to the repository and you'll to make your Microsoft and Azure organization memberships public.  Please review the Azure SDK onboarding documentation for more information and use the Onboarding Teams channel for assistance.

- [Azure SDK onboarding](https://eng.ms/docs/products/azure-developer-experience/onboard/requirements) _(Microsoft internal)_
- [Azure SDK onboarding assistance](https://teams.microsoft.com/l/channel/19%3ade3fce22c2994be18cac50502c55f717%40thread.skype/Azure%2520SDK%2520Onboarding?groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47) _(Microsoft internal)_

Please also be sure to add yourself to [CODEOWNERS](https://github.com/Azure/azure-sdk-for-net/blob/main/.github/CODEOWNERS) for this library, if you will be maintaining it going forward.

**Comment by destinyboyer** [comment]
> @destinyboyer:
> 
> Your account is missing the necessary write permissions to the repository and you'll to make your Microsoft and Azure organization memberships public. Please review the Azure SDK onboarding documentation for more information and use the Onboarding Teams channel for assistance.
> 
> * [Azure SDK onboarding](https://eng.ms/docs/products/azure-developer-experience/onboard/requirements) _(Microsoft internal)_
> * [Azure SDK onboarding assistance](https://teams.microsoft.com/l/channel/19%3ade3fce22c2994be18cac50502c55f717%40thread.skype/Azure%2520SDK%2520Onboarding?groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47) _(Microsoft internal)_
> 
> Please also be sure to add yourself to [CODEOWNERS](https://github.com/Azure/azure-sdk-for-net/blob/main/.github/CODEOWNERS) for this library, if you will be maintaining it going forward.

@jsquire I think all the necessary changes have been made.

**Comment by destinyboyer** [comment]
Access etc. should be granted. Per chatting with Scott I have joined the Azure SDK Partners SG.

**Comment by jsquire** [comment]
> Access etc. should be granted. Per chatting with Scott I have joined the Azure SDK Partners SG.

Thanks!  I'm seeing the orgs as public, but permissions not yet granted.  I'll keep an eye on them and unblock once they validate.

File: .github/CODEOWNERS
Comment (jsquire) [inline .github/CODEOWNERS:143]: As this is a management package, you'll want to put this down in the management section around [L930](https://github.com/Azure/azure-sdk-for-net/blob/main/.github/CODEOWNERS#L930) and follow the formatting there.   This should look something like:

```text
# PRLabel: %Carbon Optimization
/sdk/carbon/Azure.ResourceManager.CarbonOptimization                  @destinyboyer @liangchenmicrosoft

# ServiceLabel: %Carbon Optimization %Mgmt
# ServiceOwners:                                                      @destinyboyer
```

I'll create the service label as "Carbon Optimization"

**Comment by destinyboyer** [comment]
I expect codeowners linting to pass after the next extraction pipeline run.

**Comment by jsquire** [comment]
> I expect codeowners linting to pass after the next extraction pipeline run.

Right now, the GH API still sees you as having only read permissions:

```text
Permissions API Response:
VERBOSE:        {"permission":"read","user":{"login":"destinyboyer","id":6730144,"node_id":"MDQ6VXNlcjY3MzAxNDQ=","avatar_url":"https://avatars.githubusercontent.com/u/6730144?v=4","gravatar_id":"","url":"https://api.github.com/users/destinyboyer","html_url":"https://github.com/destinyboyer","followers_url":"https://api.github.com/users/destinyboyer/followers","following_url":"https://api.github.com/users/destinyboyer/following{/other_user}","gists_url":"https://api.github.com/users/destinyboyer/gists{/gist_id}","starred_url":"https://api.github.com/users/destinyboyer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/destinyboyer/subscriptions","organizations_url":"https://api.github.com/users/destinyboyer/orgs","repos_url":"https://api.github.com/users/destinyboyer/repos","events_url":"https://api.github.com/users/destinyboyer/events{/privacy}","received_events_url":"https://api.github.com/users/destinyboyer/received_events","type":"User","site_admin":false,"permissions":{"admin":false,"maintain":false,"push":false,"triage":false,"pull":true},"role_name":"read"},"role_name":"read"}
```

**Comment by destinyboyer** [comment]
> > I expect codeowners linting to pass after the next extraction pipeline run.
> 
> Right now, the GH API still sees you as having only read permissions:
> 
> ```
> Permissions API Response:
> VERBOSE:        {"permission":"read","user":{"login":"destinyboyer","id":6730144,"node_id":"MDQ6VXNlcjY3MzAxNDQ=","avatar_url":"https://avatars.githubusercontent.com/u/6730144?v=4","gravatar_id":"","url":"https://api.github.com/users/destinyboyer","html_url":"https://github.com/destinyboyer","followers_url":"https://api.github.com/users/destinyboyer/followers","following_url":"https://api.github.com/users/destinyboyer/following{/other_user}","gists_url":"https://api.github.com/users/destinyboyer/gists{/gist_id}","starred_url":"https://api.github.com/users/destinyboyer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/destinyboyer/subscriptions","organizations_url":"https://api.github.com/users/destinyboyer/orgs","repos_url":"https://api.github.com/users/destinyboyer/repos","events_url":"https://api.github.com/users/destinyboyer/events{/privacy}","received_events_url":"https://api.github.com/users/destinyboyer/received_events","type":"User","site_admin":false,"permissions":{"admin":false,"maintain":false,"push":false,"triage":false,"pull":true},"role_name":"read"},"role_name":"read"}
> ```

Jesse, I see that I am part of Azure SDK Partners (see screenshot below), although I do not see that getting propagated to the [github group](https://github.com/orgs/Azure/teams/azure-sdk-partners?query=). I am hoping that there might just be increased latency to get perms everywhere.

Screenshot 2024-03-25 at 3 30 15 PM

If there are any steps I am missing let me know. Is Scott or the Teams channel the best place to reach out if I don't have perms by tomorrow?

**Comment by destinyboyer** [comment]
> > > I expect codeowners linting to pass after the next extraction pipeline run.
> > 
> > 
> > Right now, the GH API still sees you as having only read permissions:
> > ```
> > Permissions API Response:
> > VERBOSE:        {"permission":"read","user":{"login":"destinyboyer","id":6730144,"node_id":"MDQ6VXNlcjY3MzAxNDQ=","avatar_url":"https://avatars.githubusercontent.com/u/6730144?v=4","gravatar_id":"","url":"https://api.github.com/users/destinyboyer","html_url":"https://github.com/destinyboyer","followers_url":"https://api.github.com/users/destinyboyer/followers","following_url":"https://api.github.com/users/destinyboyer/following{/other_user}","gists_url":"https://api.github.com/users/destinyboyer/gists{/gist_id}","starred_url":"https://api.github.com/users/destinyboyer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/destinyboyer/subscriptions","organizations_url":"https://api.github.com/users/destinyboyer/orgs","repos_url":"https://api.github.com/users/destinyboyer/repos","events_url":"https://api.github.com/users/destinyboyer/events{/privacy}","received_events_url":"https://api.github.com/users/destinyboyer/received_events","type":"User","site_admin":false,"permissions":{"admin":false,"maintain":false,"push":false,"triage":false,"pull":true},"role_name":"read"},"role_name":"read"}
> > ```
> 
> Jesse, I see that I am part of Azure SDK Partners (see screenshot below), although I do not see that getting propagated to the [github group](https://github.com/orgs/Azure/teams/azure-sdk-partners?query=). I am hoping that there might just be increased latency to get perms everywhere.
> 
> Screenshot 2024-03-25 at 3 30 15 PM
> 
> If there are any steps I am missing let me know. Is Scott or the Teams channel the best place to reach out if I don't have perms by tomorrow?

I have emailed Scott to see if there are steps I am missing. Based on the documentation I think I should have access but it doesn't seem to be propagating to GitHub.

File: .github/CODEOWNERS
Comment (destinyboyer) [reply .github/CODEOWNERS]: Done.

**Comment by destinyboyer** [comment]
Per Scott it looks like there is an outage in syncing perms. He is working with the owning team and will update me once things are in a good state.

**Comment by jsquire** [comment]
> Per Scott it looks like there is an outage in syncing perms. He is working with the owning team and will update me once things are in a good state.

If you're curious in the meantime, you can check permissions using [Valid-AzsdkCodeOwner.ps1](https://github.com/Azure/azure-sdk-tools/blob/main/tools/github/scripts/Validate-AzsdkCodeOwner.ps1)

File: .github/CODEOWNERS
Comment (jsquire) [inline .github/CODEOWNERS:931]:
```suggestion
/sdk/carbon/Azure.ResourceManager.CarbonOptimization/               @destinyboyer @liangchenmicrosoft
```

nit:  Alignment

File: .github/CODEOWNERS
Comment (jsquire) [inline .github/CODEOWNERS:934]:
```suggestion
# ServiceOwners:                                                    @destinyboyer
```

nit: Alignment

File: .github/CODEOWNERS
Comment (jsquire) [inline .github/CODEOWNERS:931]: `liangchenmicrosoft` also need write access to the repository.

Comment (destinyboyer) [reply .github/CODEOWNERS]: He has been accepted into the SG as of this morning. Once the pipelines that replicate permissions to from GitHub to CoreIdentity are unblocked I expect everything to be good.

File: sdk/carbon/Azure.ResourceManager.CarbonOptimization/README.md
Comment (ArthurMa1978) [inline sdk/carbon/Azure.ResourceManager.CarbonOptimization/README.md:3]: Add description here

File: sdk/carbon/Azure.ResourceManager.CarbonOptimization/CHANGELOG.md
Comment (ArthurMa1978) [inline sdk/carbon/Azure.ResourceManager.CarbonOptimization/CHANGELOG.md:3]: Specify the release date by replacing `Unreleased`, the format is `yyyy-mm-dd`

File: sdk/ci.mgmt.yml
Comment (ArthurMa1978) [inline sdk/ci.mgmt.yml:3]: remove this file

File: sdk/carbon/ci.mgmt.yml
Comment (ArthurMa1978) [inline sdk/carbon/ci.mgmt.yml:13]: sdk/carbon/ci.mgmt.yml

File: sdk/carbon/ci.mgmt.yml
Comment (ArthurMa1978) [inline sdk/carbon/ci.mgmt.yml:14]: sdk/carbon/Azure.ResourceManager.CarbonOptimization/

File: sdk/carbon/ci.mgmt.yml
Comment (ArthurMa1978) [inline sdk/carbon/ci.mgmt.yml:19]: carbon

File: sdk/carbon/Azure.ResourceManager.CarbonOptimization/src/Properties/AssemblyInfo.cs
Comment (ArthurMa1978) [inline sdk/carbon/Azure.ResourceManager.CarbonOptimization/src/Properties/AssemblyInfo.cs:6]: The value here is `Microsoft.Carbon`

**Comment by ArthurMa1978** [comment]
Hey @destinyboyer , I have fixed all naming errors, and please follow the [guideline](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/resourcemanager/Azure.ResourceManager/docs/TestGuide.md) to record test.

**Comment by destinyboyer** [comment]
@ArthurMa1978 when I try to run tests I am getting the following. I assume that this means there is something wrong with my specs over in the specs repo, how can I debug this?

I pulled in the merge you performed, hard reset, and regenerated files. I can see I am generating the sdk for `2024-02-01-preview`.

The `2023-04-01-preview` package is not even in the azure-rest-api-specs repo. At one point I did have tests passing locally.

```
    Azure.RequestFailedException : The resource type 'carbonEmissionReports' could not be found in the namespace 'Microsoft.Carbon' for api version '2024-02-01-preview'. The supported api-versions are '2023-04-01-preview'.
Status: 404 (Not Found)
ErrorCode: InvalidResourceType
```

**Comment by destinyboyer** [comment]
@ArthurMa1978 I am unable to record tests due to the following. I have searched the channels and see some relevant posts but none of them have fixed my issue. I am not sure if the api exception should be ignored for the inner exception or not.

The 2024-02-01-preview version is the only sdk version we have in the specs repo. Can you please help? At this point I don't know if we will make the April 3rd cutoff.

```
  Message: 
The [RecordedTest] attribute attempted to re-record, but failed:

Azure.RequestFailedException : The resource type 'queryCarbonEmissionDataAvailableDateRange' could not be found in the namespace 'Microsoft.Carbon' for api version '2024-02-01-preview'. The supported api-versions are '2023-04-01-preview'.
Status: 404 (Not Found)
ErrorCode: InvalidResourceType

Content:
{
  "error": {
    "code": "InvalidResourceType",
    "message": "The resource type 'queryCarbonEmissionDataAvailableDateRange' could not be found in the namespace 'Microsoft.Carbon' for api version '2024-02-01-preview'. The supported api-versions are '2023-04-01-preview'."
  }
}

Headers:
Date: Fri, 29 Mar 2024 16:52:56 GMT
Server: Kestrel
Cache-Control: no-cache
Pragma: no-cache
x-ms-failure-cause: REDACTED
x-ms-ratelimit-remaining-tenant-writes: REDACTED
x-ms-request-id: c2b4ca0d-229d-4930-89a5-fa843c910f22
x-ms-correlation-request-id: REDACTED
x-ms-routing-request-id: REDACTED
Strict-Transport-Security: REDACTED
X-Content-Type-Options: REDACTED
X-Cache: REDACTED
X-MSEdge-Ref: REDACTED
Content-Length: 260
Content-Type: application/json; charset=utf-8
Expires: -1


Original failure:

Azure.Core.TestFramework.TestRecordingMismatchException : The record session does not exist or is missing the Variables section. If the test is attributed with 'RecordedTest', it will be recorded automatically. Otherwise, set the RecordedTestMode to 'Record' and attempt to record the test.
```

**Comment by destinyboyer** [comment]
> @ArthurMa1978 I am unable to record tests due to the following. I have searched the channels and see some relevant posts but none of them have fixed my issue. I am not sure if the api exception should be ignored for the inner exception or not.
> 
> The 2024-02-01-preview version is the only sdk version we have in the specs repo. Can you please help? At this point I don't know if we will make the April 3rd cutoff.
> 
> ```
>   Message: 
> The [RecordedTest] attribute attempted to re-record, but failed:
> 
> Azure.RequestFailedException : The resource type 'queryCarbonEmissionDataAvailableDateRange' could not be found in the namespace 'Microsoft.Carbon' for api version '2024-02-01-preview'. The supported api-versions are '2023-04-01-preview'.
> Status: 404 (Not Found)
> ErrorCode: InvalidResourceType
> 
> Content:
> {
>   "error": {
>     "code": "InvalidResourceType",
>     "message": "The resource type 'queryCarbonEmissionDataAvailableDateRange' could not be found in the namespace 'Microsoft.Carbon' for api version '2024-02-01-preview'. The supported api-versions are '2023-04-01-preview'."
>   }
> }
> 
> Headers:
> Date: Fri, 29 Mar 2024 16:52:56 GMT
> Server: Kestrel
> Cache-Control: no-cache
> Pragma: no-cache
> x-ms-failure-cause: REDACTED
> x-ms-ratelimit-remaining-tenant-writes: REDACTED
> x-ms-request-id: c2b4ca0d-229d-4930-89a5-fa843c910f22
> x-ms-correlation-request-id: REDACTED
> x-ms-routing-request-id: REDACTED
> Strict-Transport-Security: REDACTED
> X-Content-Type-Options: REDACTED
> X-Cache: REDACTED
> X-MSEdge-Ref: REDACTED
> Content-Length: 260
> Content-Type: application/json; charset=utf-8
> Expires: -1
> 
> 
> Original failure:
> 
> Azure.Core.TestFramework.TestRecordingMismatchException : The record session does not exist or is missing the Variables section. If the test is attributed with 'RecordedTest', it will be recorded automatically. Otherwise, set the RecordedTestMode to 'Record' and attempt to record the test.
> ```

We need to put a hold on this PR. I just learned that the 2024-02-01 API version should not have been published into the specs repo yet.

**Comment by destinyboyer** [comment]
Closing per conversation with LaDonna. Will open a new PR with a new api version.

**Pull request closed without merging**
